### PR TITLE
feat: polygon l2 checkout

### PR DESCRIPTION
**Pull request by chibie**
##### Description

This PR introduces grants L2 checkout using Polygon (MATIC)

- [x] metamask switch network
- [x] metamask add polygon chain
- [x] check balance sufficiency
- [x] token allowance approval
- [x] handle bulkcheckout
- [x] post contribution to database
- [x] manual ingestion
- [x] checkout finalization
- [x] gas cost estimation
- [x] get addresses of mapped tokens (rootToChildToken)
- [x] getTokenByName
- [x] update data-chains.js
- [x] checkout recommendation
- [x] implement modal design

##### Testing

checkout flow with user having sufficient L2 balance -> https://www.loom.com/share/8506e7795b7a44dea68782478651d436

checkout flow with insufficient balance -> https://www.loom.com/share/7279ef0a89284a68b51d066f2cb005ef

handling unsupported tokens -> https://www.loom.com/share/05a5f0fe6507495ea3bbf41517663f5f

### CODE REVIEW

**Review by PixelantDesign — COMMENTED** [review]
Hi @chibie lookin' great! Thanks for creating separate videos for different scenarios. Very helpful when reviewing!

**Scenario: Checkout flow with user having sufficient L2 balance**
The confirmation says "Contribution Saved", could we consider updating the message to be along the lines of the Grants have been funded successfully message? Though I understand there is more processing that needs to happen.

Screen Shot 2021-08-30 at 8 08 49 PM

**Scenario: Checkout flow with user having sufficient L2 balance - Polygon Tooltip**
There is a lot to read in the tooltip. Could we consider simplifying by revising to the following:
Polygon is a solution that enables low cost transaction fees and transfer of tokens with fast confirmation time. To take advantage of these benefits, this method of payment requires preloaded funds in Polygon.

Screen Shot 2021-08-30 at 8 10 45 PM

**Scenario: Complete Fund Deposit Modal**

Could we consider switching how much do I have to deposit with the close modal info?
We likely want the user to focus on what they need to do next prior to thinking about closing the modal.

Screen Shot 2021-08-30 at 8 14 17 PM

**Scenario: Handling Unsupported Tokens**
We are using the save on gas costs area as an alert in the screenshot below.
Could we consider moving that information and displaying it when the user hovers over each grayed out button? 

Screen Shot 2021-08-30 at 8 24 49 PM

**Checkout Button Spacing**
Could we consider adding some space between the checkout (buttons) options? In the videos, the buttons are all connected with each other. 

Screen Shot 2021-08-30 at 9 02 39 PM

**Nice to Have**

- Determine which option would be best for the user (zksync, polygon, standard) gas wise and use purple to denote the recommendation.
- Display the amount that the user has in polygon or zksync on the button itself? This is something users mentioned would be very helpful. 
- Add service logos to each checkout button (zksync and polygon)

File: app/assets/v2/js/cart-ethereum-polygon.js [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+        }
+      });
+
+      // Compare amounts needed to balance
+      const web3 = this.initWeb3();
+      const userAddress = ethereum.selectedAddress;
+      let isBalanceSufficient = true;
+
+      for (let i = 0; i < this.cart.tokenList.length; i += 1) {
+        let tokenSymbol = this.cart.tokenList[i];
```
Comment (thelostone-mc) [inline app/assets/v2/js/cart-ethereum-polygon.js:393]: const

File: app/assets/v2/js/cart-ethereum-polygon.js [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+      // Get total amount needed for eack token by summing over donation inputs
+      this.donationInputs.forEach((donation) => {
+        const tokenSymbol = donation.name;
+        const amount = toBigNumber(donation.amount);
+
+        if (!requiredAmounts[tokenSymbol]) {
+          // First time seeing this token, set the field and initial value
+          requiredAmounts[tokenSymbol] = { amount };
+        } else {
+          // We've seen this token, so just update the total
```
Comment (thelostone-mc) [inline app/assets/v2/js/cart-ethereum-polygon.js:382]: i don't get this 😓 
> We've seen this token, so just update the total
maybe detail it a bit

File: app/assets/v2/js/cart-ethereum-polygon.js [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        const tokenAddr = currentValue.token?.toLowerCase();
+
+        if (currentValue.token === MATIC_ADDRESS) {
+          return accumulator + 70000; // MATIC donation gas estimate
+
+        } else if (tokenAddr === '0x960b236A07cf122663c4303350609A66A7B288C0'.toLowerCase()) {
+          return accumulator + 170000; // ANT donation gas estimate
+        }
+
+        return accumulator + 70000; // generic token donation gas estimate
```
Comment (thelostone-mc) [inline app/assets/v2/js/cart-ethereum-polygon.js:363]: do we wanna `70000` as a const and that there here

File: app/assets/v2/js/cart-ethereum-polygon.js [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        // TODO: find a suitable curve using
+        // https://github.com/mds1/Gitcoin-Checkout-Gas-Analysis
+        // return 27500 * donationCurrencies.length + 125000;
+        return 10000 * donationCurrencies.length + 70000;
+      }
+
+      // Otherwise, based on contract tests, we use the more conservative heuristic below to get
+      // a gas estimate. The estimates used here are based on testing the cost of a single
+      // donation (i.e. one item in the cart). Because gas prices go down with batched
+      // transactions, whereas this assumes they're constant, this gives us a conservative estimate
```
Comment (thelostone-mc) [inline app/assets/v2/js/cart-ethereum-polygon.js:352]: use multiline comment instead ?
```
/**
 *
 */

File: app/assets/v2/js/cart-ethereum-polygon.js [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+     * @dev List of tokens supported by Polygon + Gitcoin. To add a token to this list:
+     *   1. Make sure the token is top 10 used tokens based on Gitcoin's historical data
+     *   2. Confirm the token exists on Polygon's list of supported tokens: https://mapper.matic.today/
+     *   2. Add the token symbol to the appropriate list below
+     * @dev We hardcode the list from Gitcoin's historical data based on the top ten tokens
+     *   on ethereum chain and also Polygon network used by users to checkout
+     */
+    supportedTokens() {
+      const mainnetTokens = [ 'DAI', 'ETH', 'USDT', 'USDC', 'PAN', 'BNB', 'UNI', 'CELO', 'MASK', 'MATIC' ];
+      const testnetTokens = [ 'MATIC', 'ETH', 'DAI' ];
```
Comment (thelostone-mc) [inline app/assets/v2/js/cart-ethereum-polygon.js:63]: ^ do we wanna fetch these from DB ? we can use https://gitcoin.co/_administrationperftools/staticjsonenv/

**Review by thelostone-mc — COMMENTED** [review]
left a few comments

**Comment by willsputra** [comment]
Looking good @chibie !

Could we consider showing the next steps immediately after switching network to Polygon?

Current flow:
Click "Checkout with Polygon" -> Switch Network -> Click "Checkout with Polygon" again -> payout/insufficient fund modal

Ideal flow:
Click "Checkout with Polygon" -> Switch Network -> payout/insufficient fund modal
so users won't need to click "Checkout with Polygon" the second time

File: app/assets/v2/js/cart-ethereum-polygon.js (hunk 3 above)
Comment (chibie) [reply app/assets/v2/js/cart-ethereum-polygon.js:369]: the gas estimation work isn't accurate yet... more work on this in GITC-370 + GITC-372

File: app/assets/v2/js/cart-ethereum-polygon.js (hunk 5 above)
Comment (chibie) [reply app/assets/v2/js/cart-ethereum-polygon.js:63]: isn't that an extra http request? (--performance) => we won't have to change these in a long time

File: app/assets/v2/js/cart-ethereum-polygon.js (hunk 2 above)
Comment (chibie) [reply app/assets/v2/js/cart-ethereum-polygon.js:388]: haha... i'll reword it

**Comment by chibie** [comment]
i've attended to the review comments except the following from @PixelantDesign above

* Scenario: Handling Unsupported Tokens
* Checkout Button Spacing

these should be handled with the checkout page refresh (separate ticket)

**Review by thelostone-mc — APPROVED** [review]
MERGING THIS INTO GR11. 
Please create a ticket and drop in feedback

**Pull request merged**